### PR TITLE
clst-proto initial setup

### PR DESCRIPTION
**Pull request by ottaviohartman**

### CODE REVIEW

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/serde/RecordSerDe.java [hunk 1]
```diff
@@ -28,8 +31,10 @@
     String valueSchemaId;
   }
 
+  // we must implement this method & reach out to Buf
```
Comment (ottaviohartman) [inline kafka-ui-api/src/main/java/com/provectus/kafka/ui/serde/RecordSerDe.java]: TODO: remove these

File: kafka-ui-api/src/main/resources/application-local.yml [hunk 2]
```diff
@@ -4,11 +4,12 @@ kafka:
       bootstrapServers: localhost:9092
       zookeeper: localhost:2181
       schemaRegistry: http://localhost:8085
-      ksqldbServer: http://localhost:8088
+      # ksqldbServer: http://localhost:8088
```
Comment (ottaviohartman) [inline kafka-ui-api/src/main/resources/application-local.yml]: Undo these changes, make a new file

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/serde/schemaregistry/BufAndSchemaRegistryAwareRecordSerDe.java [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+  public ProducerRecord<byte[], byte[]> serialize(String topic,
+                                                  @Nullable String key,
+                                                  @Nullable String data,
+                                                  @Nullable Integer partition) {
+    return this.schemaRegistryAwareRecordSerDe.serialize(topic, key, data, partition);
+  }
+
+  @Override
+  public TopicMessageSchemaDTO getTopicSchema(String topic) {
+    // TODO
```
Comment (ottaviohartman) [inline kafka-ui-api/src/main/java/com/provectus/kafka/ui/serde/schemaregistry/BufAndSchemaRegistryAwareRecordSerDe.java]: TODO

File: .github/clear-street/README.md [hunk 4]
```diff
@@ -0,0 +1,66 @@
+# Clear Street Proto with Buf
+
+## Goal
+
+At Clear Street, protobuf on Kafka is enabled with some special Kafka Headers or topic names.
+We've added [Buf](https://buf.build/) support to Kafka UI, as well as special deserialization logic.
+
+Use the [Protobuf on Kafka](https://www.notion.so/clearstreet/Protobuf-on-Kafka-ac912401cf504c75956d907b86a12f55) to reference the format.
```
Comment (Vishesh-Gupta) [inline .github/clear-street/README.md:8]: this is an internal document though and this repo is public. I don't think it is a good idea to link something that others can't access in a public README.md

Comment (ottaviohartman) [reply .github/clear-street/README.md:8]: I think it's safe; it's behind a Notion login

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/serde/schemaregistry/BufAndSchemaRegistryAwareRecordSerDe.java [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+    if (protoSchema != null) {
+      // TODO: parse message and return it
+      log.info("Skipping buf for schema: {}", protoSchema.fullyQualifiedTypeName);
+      return this.schemaRegistryAwareRecordSerDe.deserialize(msg);
+    }
+
+    return this.schemaRegistryAwareRecordSerDe.deserialize(msg);
+  }
+
+  private @Nullable ProtoSchema protoSchemaFromHeaders(Headers headers) {
```
Comment (jdechicchis) [inline kafka-ui-api/src/main/java/com/provectus/kafka/ui/serde/schemaregistry/BufAndSchemaRegistryAwareRecordSerDe.java]: We should probably add units test around this (maybe just scaffold it and have it pass for now). Is it easy to add with how kafkaui is set up?

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/serde/schemaregistry/BufAndSchemaRegistryAwareRecordSerDe.java [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+      schemaID = new String(header.value());
+    }
+
+    ProtoSchema ret = new ProtoSchema();
+    ret.fullyQualifiedTypeName = fullyQualifiedTypeName;
+    ret.schemaID = schemaID;
+    return ret;
+  }
+
+  private @Nullable ProtoSchema protoSchemaFromTopic(String topic) {
```
Comment (jdechicchis) [inline kafka-ui-api/src/main/java/com/provectus/kafka/ui/serde/schemaregistry/BufAndSchemaRegistryAwareRecordSerDe.java]: Same thing about unit tests here

**Review by jdechicchis — APPROVED** [review]
These are more curiosities. We can tackle unit tests later

File: .github/clear-street/README.md (hunk 4 above)
Comment (jdechicchis) [reply .github/clear-street/README.md:8]: Agree for now. Eventually, since this is open source/if we want to merge upstream we'll want to clean it up

Comment (Vishesh-Gupta) [reply .github/clear-street/README.md:8]: yeah it is behind a Notion login, but I was against having something internal here. It is just a format, we can mention that here. Copy paste it from the doc. Kind of cleans up in advance if we want in upstream or not.

Comment (ottaviohartman) [reply .github/clear-street/README.md:8]: That leaves two places to update the reference; I'll keep it here for now

File: kafka-ui-api/src/main/java/com/provectus/kafka/ui/serde/schemaregistry/BufAndSchemaRegistryAwareRecordSerDe.java (hunk 5 above)
Comment (ottaviohartman) [reply kafka-ui-api/src/main/java/com/provectus/kafka/ui/serde/schemaregistry/BufAndSchemaRegistryAwareRecordSerDe.java]: Ah, I meant to. I'll do that now

File: kafka-ui-api/src/test/java/com/provectus/kafka/ui/serde/schemaregistry/BufAndSchemaRegistryAwareRecordSerDeTest.java [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+    @Test
+    void callsBufFormatterWhenTopicIsCorrect() throws Exception {
+      ConsumerRecord<Bytes, Bytes> record =  new ConsumerRecord<>(
+          "test-topic.proto.foo",
+          1,
+          100,
+          Bytes.wrap("key".getBytes()),
+          Bytes.wrap("value".getBytes())
+      );
+      record.headers().add("PROTOBUF_TYPE", "protobuf_type".getBytes());
```
Comment (jdechicchis) [inline kafka-ui-api/src/test/java/com/provectus/kafka/ui/serde/schemaregistry/BufAndSchemaRegistryAwareRecordSerDeTest.java]: I think we should remove this line so it tests that Buf is used if the topic is set even though the header isn't set

Comment (ottaviohartman) [reply kafka-ui-api/src/test/java/com/provectus/kafka/ui/serde/schemaregistry/BufAndSchemaRegistryAwareRecordSerDeTest.java]: good catch! Fixed
